compiler/z/codegen/UnaryEvaluator.cpp

```diff
@@ -485,19 +485,21 @@ OMR::Z::TreeEvaluator::l2aEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    TR_ASSERT(hasCompPtrs, "no compression sequence found under l2a node [%p]\n", node);
    */
 
-   // TODO (GuardedStorage)
-   TR::Register *source;
-   if (comp->useCompressedPointers() && (TR::Compiler->om.compressedReferenceShift() == 0 || firstChild->containsCompressionSequence()) && !node->isl2aForCompressedArrayletLeafLoad())
-      cg->setEvalCompressionSequence(true);
+   TR::Register* source = NULL;
 
-   if ((TR::Compiler->target.cpu.getS390SupportsGuardedStorageFacility()) && comp->useCompressedPointers() && (firstChild->containsCompressionSequence()) && !node->isl2aForCompressedArrayletLeafLoad())
+   // TODO (GuardedStorage): Why is isl2aForCompressedArrayletLeafLoad check necessary?
```

Comment:
As for the other part of the question I do not think this can currently happen. All compression sequences are anchored under `TR::compressedRefs` tree anchors, and following the compressedrefs rematerialization phase they are expanded into the compression / decompression sequences. 

Though I do share your concern in that something like lower trees could decide to sink a reference down to a grandchild in which case we would encounter the scenario you speak of. This problem is however avoided with a pre-instruction selection pass which anchors all `TR::l2a` trees to their own treetops. See:

https://github.com/eclipse/omr/blob/master/compiler/codegen/OMRCodeGenPhase.cpp#L423-L454

This whole idea needs to be reworked. I do not think we should be rematerializing `TR::compressedRefs` trees to begin with and avoiding to generate the compression sequence. The guarded load instruction models the load and compression together, and as such we should not explicitly represent the compression in the trees. This is work that I will be doing next.